src/disk_cache.rs

```diff
@@ -0,0 +1,955 @@
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum InvalidGetRange {
+    #[error("Range start too large, requested: {requested}, length: {length}")]
+    StartTooLarge { requested: usize, length: usize },
+
+    #[error("Range started at {start} and ended at {end}")]
+    Inconsistent { start: usize, end: usize },
+}
+
+struct DiskCacheEntry {
+    root_folder: std::path::PathBuf,
+    location: object_store::path::Path,
+    part_size: usize,
+}
+
+#[allow(unused)]
+type PartID = usize;
+
+#[allow(unused)]
+impl DiskCacheEntry {
+    /// save the GetResult to the disk cache, a GetResult may be transformed into multiple part
+    /// files and a meta file. please note that the `range` in the GetResult is expected to be
+    /// aligned with the part size.
+    pub async fn save_result(&self, result: GetResult) -> object_store::Result<usize> {
+        assert!(result.range.start % self.part_size == 0);
+        assert!(result.range.end % self.part_size == 0 || result.range.end == result.meta.size);
+
+        self.save_meta(&result.meta).await?;
+
+        let mut buffer = BytesMut::new();
+        let mut part_number = result.range.start / self.part_size;
+        let object_size = result.meta.size;
+
+        let mut stream = result.into_stream();
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk?;
+            buffer.extend_from_slice(&chunk);
+
+            while buffer.len() >= self.part_size {
+                let to_write = buffer.split_to(self.part_size);
+                self.save_part(part_number, to_write.as_ref()).await?;
+                part_number += 1;
+            }
+        }
+
+        // if the last part is not fully filled, save it as the last part.
+        if !buffer.is_empty() {
+            self.save_part(part_number, buffer.as_ref()).await?;
+            return Ok(object_size);
+        }
+
+        Ok(object_size)
+    }
+
+    pub async fn read_part(&self, part_id: PartID) -> object_store::Result<Option<Bytes>> {
+        let part_path = self.make_part_path(part_id, false);
+
+        // if the part file does not exist, return None
+        let exists = fs::try_exists(&part_path).await.unwrap_or(false);
+        if !exists {
+            return Ok(None);
+        }
+
+        let file = File::open(&part_path).await.map_err(wrap_io_err)?;
+        let mut reader = tokio::io::BufReader::new(file);
+        let mut buffer = Vec::new();
+        reader.read_to_end(&mut buffer).await.map_err(wrap_io_err)?;
+        Ok(Some(Bytes::from(buffer)))
+    }
+
+    pub async fn read_meta(&self) -> object_store::Result<Option<ObjectMeta>> {
+        let meta_path = self.make_meta_path();
+
+        let exists = fs::try_exists(&meta_path).await.unwrap_or(false);
+        if !exists {
+            return Ok(None);
+        }
+
+        // TODO: process not found err here instead of check exists
+        let file = File::open(&meta_path).await.map_err(wrap_io_err)?;
```

Comment:
this reminds me an article from sqlite3 https://www.sqlite.org/fasterthanfs.html , it said that storing files in sqlite gives a performance gain of 35% because of all the data access in sqlite3 is in a single file descriptor without the cost of duplicated opening/closing.

i think we can make a trait about the local cache storage, user can even choose storing their local cache files in sqlite3 :)